File: android/app/src/main/java/com/secondscreen/local/update/Updater.kt
package com.secondscreen.local.update

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

data class ReleaseInfo(
    val available: Boolean,
    val latestVersion: String,
    val currentVersion: String,
    val apkUrl: String?,
    val notes: String,
    val message: String
)

// Checks GitHub Releases for a newer APK and installs it. Internet is used ONLY here (on demand);
// everything else in the app is offline/LAN.
object Updater {
    private const val REPO_OWNER = "fajarmaola"
    private const val REPO_NAME = "HpToMonitor"

    suspend fun check(currentVersion: String): ReleaseInfo = withContext(Dispatchers.IO) {
        try {
            val api = "https://api.github.com/repos/$REPO_OWNER/$REPO_NAME/releases/latest"
            val json = httpGet(api) ?: return@withContext fail(currentVersion, "Tidak dapat menghubungi GitHub.")
            val root = JSONObject(json)
            var apkUrl: String? = null
            var versionJsonUrl: String? = null
            val assets = root.optJSONArray("assets")
            if (assets != null) {
                for (i in 0 until assets.length()) {
                    val a = assets.getJSONObject(i)
                    val name = a.optString("name")
                    val dl = a.optString("browser_download_url")
                    if (name.endsWith(".apk", true)) apkUrl = dl
                    else if (name.equals("version.json", true)) versionJsonUrl = dl
                }
            }
            var latest = root.optString("tag_name", "")
            var notes = ""
            if (versionJsonUrl != null) {
                val vj = httpGet(versionJsonUrl)
                if (vj != null) {
                    val vo = JSONObject(vj)
                    latest = vo.optString("version", latest)
                    notes = vo.optString("notes", "")
                }
            }
            latest = latest.trimStart('v', 'V')
            val available = compare(latest, currentVersion) > 0 && apkUrl != null
            ReleaseInfo(available, latest, currentVersion, apkUrl, notes,
                if (apkUrl == null) "Aset APK tidak ditemukan di rilis." else "")
        } catch (e: Exception) {
            fail(currentVersion, "Gagal memeriksa pembaruan: ${e.message}")
        }
    }

    suspend fun download(ctx: Context, url: String): File? = withContext(Dispatchers.IO) {
        try {
            val out = File(ctx.cacheDir, "HPkeMonitor-update.apk")
            if (out.exists()) out.delete()
            val conn = URL(url).openConnection() as HttpURLConnection
            conn.instanceFollowRedirects = true
            conn.connectTimeout = 20000
            conn.readTimeout = 60000
            conn.inputStream.use { input -> out.outputStream().use { input.copyTo(it) } }
            conn.disconnect()
            out
        } catch (e: Exception) { null }
    }

    fun install(ctx: Context, apk: File) {
        val uri: Uri = FileProvider.getUriForFile(ctx, ctx.packageName + ".fileprovider", apk)
        val intent = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, "application/vnd.android.package-archive")
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        ctx.startActivity(intent)
    }

    private fun fail(cur: String, msg: String) = ReleaseInfo(false, "", cur, null, "", msg)

    private fun httpGet(urlStr: String): String? {
        return try {
            val conn = URL(urlStr).openConnection() as HttpURLConnection
            conn.instanceFollowRedirects = true
            conn.connectTimeout = 15000
            conn.readTimeout = 20000
            conn.setRequestProperty("User-Agent", "HPkeMonitor-Updater")
            conn.setRequestProperty("Accept", "application/vnd.github+json")
            val code = conn.responseCode
            if (code in 200..299) conn.inputStream.bufferedReader().use { it.readText() } else null
        } catch (e: Exception) { null }
    }

    private fun compare(a: String, b: String): Int {
        val pa = parse(a); val pb = parse(b)
        for (i in 0 until 4) {
            val x = pa.getOrElse(i) { 0 }; val y = pb.getOrElse(i) { 0 }
            if (x != y) return x.compareTo(y)
        }
        return 0
    }

    private fun parse(v: String): List<Int> {
        val list = mutableListOf<Int>()
        for (p in v.split('.', '-', '+', ' ')) {
            val n = p.toIntOrNull()
            if (n != null) list.add(n) else if (list.isNotEmpty()) break
        }
        return list
    }
}

File: windows/SecondScreen.Core/Update/Updater.cs
using System.Diagnostics;
using System.Net.Http;
using System.Text.Json;

namespace SecondScreen.Core;

public sealed class UpdateInfo
{
    public bool Available;
    public string LatestVersion = "";
    public string CurrentVersion = "";
    public string? InstallerUrl;
    public string Notes = "";
    public string Message = "";
}

// Checks GitHub Releases for a newer build and downloads/launches the installer. Internet is used
// ONLY when the user taps "Check for updates" — all streaming/config stays offline (LAN).
// The release feed is the repo's "latest" release; version comes from the attached version.json.
public static class Updater
{
    // Public GitHub repo that hosts the Releases (installer + apk + version.json).
    public const string RepoOwner = "fajarmaola";
    public const string RepoName = "HpToMonitor";

    private static readonly HttpClient Http = CreateClient();

    private static HttpClient CreateClient()
    {
        var c = new HttpClient { Timeout = TimeSpan.FromSeconds(25) };
        c.DefaultRequestHeaders.UserAgent.ParseAdd("HPkeMonitor-Updater");
        c.DefaultRequestHeaders.Accept.ParseAdd("application/vnd.github+json");
        return c;
    }

    public static string CurrentVersion()
    {
        var v = System.Reflection.Assembly.GetEntryAssembly()?.GetName().Version;
        return v == null ? "0.0.0" : $"{v.Major}.{v.Minor}.{v.Build}";
    }

    public static async Task<UpdateInfo> CheckAsync()
    {
        var info = new UpdateInfo { CurrentVersion = CurrentVersion() };
        try
        {
            var url = $"https://api.github.com/repos/{RepoOwner}/{RepoName}/releases/latest";
            var json = await Http.GetStringAsync(url);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            string? versionJsonUrl = null;
            string? exeUrl = null;
            if (root.TryGetProperty("assets", out var assets))
            {
                foreach (var a in assets.EnumerateArray())
                {
                    var name = a.GetProperty("name").GetString() ?? "";
                    var dl = a.GetProperty("browser_download_url").GetString();
                    if (name.Equals("version.json", StringComparison.OrdinalIgnoreCase)) versionJsonUrl = dl;
                    else if (name.EndsWith(".exe", StringComparison.OrdinalIgnoreCase)) exeUrl = dl;
                }
            }

            string latest = root.TryGetProperty("tag_name", out var tg) ? (tg.GetString() ?? "") : "";
            if (versionJsonUrl != null)
            {
                try
                {
                    var vj = await Http.GetStringAsync(versionJsonUrl);
                    using var vd = JsonDocument.Parse(vj);
                    if (vd.RootElement.TryGetProperty("version", out var vv)) latest = vv.GetString() ?? latest;
                    if (vd.RootElement.TryGetProperty("notes", out var nn)) info.Notes = nn.GetString() ?? "";
                }
                catch { /* fall back to tag_name */ }
            }

            latest = latest.TrimStart('v', 'V');
            info.LatestVersion = latest;
            info.InstallerUrl = exeUrl;
            info.Available = Compare(latest, info.CurrentVersion) > 0 && exeUrl != null;
            if (exeUrl == null) info.Message = "Aset installer (.exe) tidak ditemukan di rilis terbaru.";
            return info;
        }
        catch (Exception ex)
        {
            info.Message = "Gagal memeriksa pembaruan: " + ex.Message;
            return info;
        }
    }

    public static async Task<string?> DownloadInstallerAsync(string url, IProgress<double>? progress = null)
    {
        try
        {
            using var resp = await Http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead);
            resp.EnsureSuccessStatusCode();
            long total = resp.Content.Headers.ContentLength ?? -1;
            var path = Path.Combine(Path.GetTempPath(), "HPkeMonitor-Setup.exe");
            await using (var src = await resp.Content.ReadAsStreamAsync())
            await using (var dst = File.Create(path))
            {
                var buf = new byte[81920];
                long read = 0; int n;
                while ((n = await src.ReadAsync(buf)) > 0)
                {
                    await dst.WriteAsync(buf.AsMemory(0, n));
                    read += n;
                    if (total > 0) progress?.Report((double)read / total);
                }
            }
            return path;
        }
        catch { return null; }
    }

    // Runs the downloaded installer SILENTLY (no wizard clicks) so an update feels like an in-app
    // update, not a manual reinstall. Inno Setup closes the running app (Restart Manager) + our own
    // Application.Shutdown frees the files, then the installer relaunches the app in silent mode.
    // A single UAC prompt is unavoidable because the install writes to Program Files + installs the
    // display driver (pnputil) — both require elevation.
    public static void RunInstaller(string path) =>
        Process.Start(new ProcessStartInfo(path)
        {
            UseShellExecute = true,
            Arguments = "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /CLOSEAPPLICATIONS"
        });

    // Returns >0 when a > b, <0 when a < b, 0 when equal (numeric, dotted).
    private static int Compare(string a, string b)
    {
        var pa = Parse(a); var pb = Parse(b);
        for (int i = 0; i < 4; i++)
        {
            int x = i < pa.Length ? pa[i] : 0;
            int y = i < pb.Length ? pb[i] : 0;
            if (x != y) return x.CompareTo(y);
        }
        return 0;
    }

    private static int[] Parse(string v)
    {
        var list = new List<int>();
        foreach (var p in v.Split('.', '-', '+', ' '))
        {
            if (int.TryParse(p, out var n)) list.Add(n);
            else if (list.Count > 0) break;
        }
        return list.ToArray();
    }
}
